### PR TITLE
Awais786/ecom 6293 list as pc

### PR DESCRIPTION
**Pull request by awais786**
PC user can see all Courses assigned to him needing Studio Instance. User should be part o f Partner coordinator group and also user should be assigned as PC for a course.
As a global admin can see all courses.

### CODE REVIEW

**Comment by awais786** [comment]
@tasawernawaz @waheedahmed Please review this PR.

File: course_discovery/apps/publisher/constants.py [hunk 1]
```diff
@@ -1,2 +1,5 @@
 # Name of the administrative group for the Publisher app
 ADMIN_GROUP_NAME = 'Publisher Admins'
+PARTNER_GROUP_NAME = 'Partner Coordinator Admins'
+REVIEWER_GROUP_NAME = 'Reviewer Admins'
+PUBLISHER_GROUP_NAME = 'publisher Admins'
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/constants.py]: These are not admins, change the names to:
```PARTNER_GROUP_NAME = 'Partner Coordinators'```
```REVIEWER_GROUP_NAME = 'Marketing Reviewers'```
```PUBLISHER_GROUP_NAME  = 'Publishers'```

File: course_discovery/apps/publisher/tests/test_views.py [hunk 2]
```diff
@@ -1019,15 +1027,37 @@ def test_page_without_login(self):
...
+        self.assertEqual(len(response.context['studio_request_courses']), 3)
+        self.assertEqual(len(response.context['published_course_runs']), 1)
+        self.assertEqual(len(response.context['in_progress_course_runs']), 0)
+        self.assertEqual(len(response.context['preview_course_runs']), 0)
+
+    def test_with_group_admin_with_in_progress(self):
+        """ Verify that admin user can see all tabs with all course runs from all groups. """
+
+        # change the course state for in-progress course run
+        self.course_run_2.change_state(target=State.NEEDS_FINAL_APPROVAL)
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/tests/test_views.py]: I think definition of in-progress is: A course-run is in all states (Draft, In-Review, Need-Approval, etc) except ```Published``` state.

File: course_discovery/apps/publisher/views.py [hunk 3]
```diff
@@ -38,13 +41,17 @@ class Dashboard(mixins.LoginRequiredMixin, ListView):
...
 
     def get_queryset(self):
-        if self.request.user.is_staff:
+        user = self.request.user
+        course_runs = None
+        if is_publisher_admin(user):
             course_runs = CourseRun.objects.select_related('course').all()
-        else:
-            courses = get_objects_for_user(self.request.user, Course.VIEW_PERMISSION, Course)
+        elif is_partner_coordinator_user(user):
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/views.py]: What about other tabs? With this code a PC can only see the course-runs which are currently assigned to him as PC in all other tabs, what if he is a assigned as marketing reviewer or publisher for some course-runs then he will not see them in in-progress or published tabs.

**Review by waheedahmed — CHANGES_REQUESTED** [review]
Right now the implementation doesn't look right to me, I think first you should re-think your implementation.

File: course_discovery/apps/publisher/tests/test_views.py [hunk 4]
```diff
@@ -1016,98 +1034,103 @@ def test_page_without_login(self):
...
-        self.course_run_2.change_state(target=State.NEEDS_FINAL_APPROVAL)
-        self.course_run_2.save()
-
-        response = self.assert_dashboard_response()
-        self.assertEqual(len(response.context['in_progress_course_runs']), 1)
+        response = self.assert_dashboard_response(studio_cnt=2, preview_cnt=1, progress_cnt=1, published_cnt=1)
         self.assertContains(response, self.table_class.format(id='in-progress'))
 
-    def assert_dashboard_response(self):
+    def assert_dashboard_response(self, studio_cnt=0, published_cnt=0, progress_cnt=0, preview_cnt=0):
```
Comment (bderusha) [inline course_discovery/apps/publisher/tests/test_views.py]: please use full word `count`

File: course_discovery/apps/publisher/tests/test_views.py [hunk 5]
```diff
@@ -1016,98 +1034,103 @@ def test_page_without_login(self):
...
-        self.assertEqual(len(response.context['preview_course_runs']), 0)
-        self.assertContains(response, 'There are no course runs marked for preview.')
-
     def test_without_preview_url(self):
         """ Verify preview ready tabs shows a message if no course run available. """
         self.course_run_2.preview_url = None
         self.course_run_2.save()
-        response = self.assert_dashboard_response()
-        self.assertEqual(len(response.context['preview_course_runs']), 0)
+        response = self.assert_dashboard_response(studio_cnt=2, preview_cnt=0, progress_cnt=1, published_cnt=1)
```
Comment (bderusha) [inline course_discovery/apps/publisher/tests/test_views.py]: please use full word `count`

File: course_discovery/apps/publisher/tests/test_views.py [hunk 6]
```diff
@@ -1016,98 +1034,103 @@ def test_page_without_login(self):
...
+        """
+        self.client.logout()
+        user = UserFactory()
+        group = factories.GroupFactory()
+        user.groups.add(group)
+
+        self.client.login(username=user.username, password=USER_PASSWORD)
+        assign_perm(Course.VIEW_PERMISSION, group, self.course_run_1.course)
+        assign_perm(Course.VIEW_PERMISSION, group, self.course_run_2.course)
+        self.assert_dashboard_response(studio_cnt=2, published_cnt=0, progress_cnt=1, preview_cnt=1)
```
Comment (bderusha) [inline course_discovery/apps/publisher/tests/test_views.py]: please use full word `count`

File: course_discovery/apps/publisher/tests/test_views.py [hunk 7]
```diff
@@ -1016,98 +1034,103 @@ def test_page_without_login(self):
...
 
-    def test_page_with_different_group_user(self):
-        """ Verify that user from one group can access only that group courses. """
+    def test_un_authorize_group_user_cannot_view_courses(self):
+        """ Verify that user from un-authorize group can access only that group courses. """
         self.client.logout()
-        self.client.login(username=self.user2.username, password=USER_PASSWORD)
-        self.assert_dashboard_response()
+        self.client.login(username=UserFactory(), password=USER_PASSWORD)
+        self.assert_dashboard_response(0, 0, 0, 0)
```
Comment (bderusha) [inline course_discovery/apps/publisher/tests/test_views.py]: Prefer naming the kwargs.  `self.assert_dashboard_response(studio_cnt=2, published_cnt=0, progress_cnt=1, preview_cnt=1)`  Much more readable that way and clear what you are testing for.

**Review by bderusha — APPROVED** [review]
A few small changes.  👍 once implemented

File: course_discovery/apps/publisher/constants.py [hunk 8]
```diff
@@ -1,2 +1,6 @@
 # Name of the administrative group for the Publisher app
 ADMIN_GROUP_NAME = 'Publisher Admins'
+PARTNER_GROUP_NAME = 'Partner Coordinators'
```
Comment (bderusha) [inline course_discovery/apps/publisher/constants.py]: `PARTNER_COORDINATOR_GROUP_NAME`

**Comment by bderusha** [comment]
@awais786 You should be able to set up a sandbox now.  FYI https://openedx.atlassian.net/wiki/display/EdxOps/Sandboxes#Sandboxes-ProvisioningwithCatalogService(CourseDiscovery)

**Review by waheedahmed — CHANGES_REQUESTED** [review]
The code looks good now, but I think you missed one check:
1. Validate that a PC can see all course-runs assigned to him/her that need a Studio Instance created.

The current code will show all course runs which require studio instance even if a user is not PC for them but he is a Reviewer or Publisher.

**Comment by awais786** [comment]
@bderusha can you please take a look again i have added the missing condition for studio requests for PC.

File: course_discovery/apps/publisher/tests/test_views.py [hunk 9]
```diff
@@ -1016,98 +1039,108 @@ def test_page_without_login(self):
...
-        self.course_run_1.lms_course_id = 'test'
-        self.course_run_1.save()
-        response = self.assert_dashboard_response()
+        admin_user = UserFactory()
+        admin_user.groups.add(self.group_admin)
+        self.client.login(username=admin_user.username, password=USER_PASSWORD)
+        self.assert_dashboard_response(studio_count=4, published_count=1, progress_count=1, preview_count=1)
+
+    def test_with_internal_group(self):
+        """ Verify that inter user can see courses assigned to the groups. """
```
Comment (bderusha) [inline course_discovery/apps/publisher/tests/test_views.py]: `internal`

File: course_discovery/apps/publisher/tests/test_views.py [hunk 10]
```diff
@@ -1016,98 +1039,108 @@ def test_page_without_login(self):
...
+        self.client.logout()
+        self.client.login(username=self.user1.username, password=USER_PASSWORD)
+        self.assert_dashboard_response(studio_count=2, published_count=1, progress_count=1, preview_count=1)
+
+    def test_with_permissions(self):
+        """ Verify that user can view only those courses on which user group have permissions assigned. """
+        self.client.logout()
+        user = UserFactory()
+        self.client.login(username=user.username, password=USER_PASSWORD)
+        assign_perm(Course.VIEW_PERMISSION, factories.GroupFactory(), self.course_run_1.course)
```
Comment (bderusha) [inline course_discovery/apps/publisher/tests/test_views.py:1109]: We should be assigning permission to the org

Comment (bderusha) [reply course_discovery/apps/publisher/tests/test_views.py:1109]: here and below

Comment (awais786) [reply course_discovery/apps/publisher/tests/test_views.py:1109]: @bderusha That is not part of this story. We have to remove the Group Permissions and fix all tests cases which should be a separate ticket.

Comment (bderusha) [reply course_discovery/apps/publisher/tests/test_views.py:1109]: gotcha 👍

File: course_discovery/apps/publisher/views.py [hunk 11]
```diff
@@ -72,7 +76,19 @@ def get_context_data(self, **kwargs):
 
         unpublished_course_runs = course_runs.exclude(state__name=State.PUBLISHED)
 
-        studio_request_courses = unpublished_course_runs.filter(lms_course_id__isnull=True)
+        # Studio requests needs to check depending upon the user role with course
+        # Also user should be part of partner coordinator group.
+
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/views.py]: Remove this line.

File: course_discovery/apps/publisher/views.py [hunk 12]
```diff
@@ -72,7 +76,19 @@ def get_context_data(self, **kwargs):
...
+        if is_publisher_admin(self.request.user):
+            studio_request_courses = unpublished_course_runs.filter(lms_course_id__isnull=True)
+
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/views.py]: Remove this line.

File: course_discovery/apps/publisher/views.py [hunk 13]
```diff
@@ -72,7 +76,19 @@ def get_context_data(self, **kwargs):
...
+        elif is_partner_coordinator_user(self.request.user):
+            studio_request_courses = unpublished_course_runs.filter(lms_course_id__isnull=True).filter(
+                course__course_user_roles__role=PublisherUserRole.PartnerCoordinator
+            )
+
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/views.py]: Remove this line.

**Review by waheedahmed — APPROVED** [review]
Some small nits otherwise 👍

**Pull request merged**